### ISSUE TITLE
Connecting to existing session in kitty terminal breaks terminal integration

### ISSUE THREAD

**Issue by joefiorini** [mawww/kakoune#4488]
### Version of Kakoune

nightly-85-g43d4276

### Reproducer

Using kitty terminal open two splits or tabs. In the first window open Kakoune. In the second window open an editor connected to the first editor's session. Then in the second window run a command that uses the terminal like `:kitty-terminal fzf` & then close it with no action. Run the same command the first terminal and close it as well. You should see the command working as you would expect. Now in the _first_ window (the one that created the session) press ctrl+z to send SIGSTOP to kakoune, you should be at a shell prompt. Use `fg` to bring kakoune back to the foreground. Run the same command that you ran above in the first (session host) window, then run it in the second window.

### Outcome

Terminal commands after I've backgrounded & foregrounded the session host using SIGSTOP do not work. They fail with a stack trace (see below) in the `*debug*` buffer.

### Expectations

I would expect the commands to work as they did before backgrounding.

### Additional information

I've tried this in multiple shells, using different programs and other terminals. This seems to be a problem to kakoune and its kitty integration. I originally opened this as kovidgoyal/kitty#4148, but the author closed it saying its a problem with kakoune. For reference here is the stack trace I get:

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/__main__.py", line 152, in <module>
    main()
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/__main__.py", line 148, in main
    func(sys.argv[1:])
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/__main__.py", line 22, in remote_control
    rc_main(args)
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/kitty/remote_control.py", line 242, in main
    response = do_io(global_opts.to, send, no_response, response_timeout)
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/kitty/remote_control.py", line 151, in do_io
    with io:
  File "/opt/src/kitty/linux-package/bin/../lib/kitty/kitty/utils.py", line 455, in __enter__
    self.tty_fd, self.original_termios = open_tty(self.read_with_timeout)
OSError: Failed to open controlling terminal: /dev/tty (identified with ctermid()) with error: No such device or address
```

**Comment by joefiorini** [comment]
I just noticed today that this occurs anytime I connect to an existing session, not just across windows. Meaning that running:

```
kak -d -s test
kak -c test
```

will also reproduce the issue.

**Comment by joefiorini** [comment]
Updated title to reflect the new findings I listed above. This is not just an issue with using SIGSTOP but a general issue when connecting to a session in kitty terminal.

**Comment by Screwtapello** [comment]
Using Kitty 0.21.1, I did the following

- launched Kitty
- pressed Ctrl-Shift-T to make a second tab
- In the first tab, I ran `TERM=xterm-kitty kak -d -s test`
- In the second tab, I ran `TERM=xterm-kitty kak -c test`, then `:b *debug*` to look at the `*debug*` buffer, no traceback
- In the second tab, I ran `:kitty-terminal-tab sleep 10`, still no traceback
- In the first tab, I put the server process into the background, then tried to quit the shell
    - The shell quit, but the terminal tab didn't close, for reasons I don't understand,
- In the second tab, I ran `:kitty-terminal-tab sleep 10` again, and this time I got the traceback

It looks like when Kakoune runs `kitty @ launch` to create a new window or tab, the new `kitty` command tries to connect to the existing terminal via `/dev/tty`, the terminal it's running inside. However, if the Kakoune server is not running inside a terminal (because it got daemonised, or the terminal it was launched from was closed), the new `kitty` command crashes.

According to #3530, if you launch Kitty with the `--listen-on` switch, then Kitty will listen for commands via `/dev/tty` *and* via the method named in the switch. That method is exposed to Kakoune via the `$KITTY_LISTEN_ON`environment variable, and if it's present then the Kakoune server doesn't need `/dev/tty` and the Kitty integration should work just fine regardless of whether the server is inside a terminal or not.
